src/content/en/updates/posts/2016/03/access-usb-devices-on-the-web.markdown

```diff
@@ -0,0 +1,280 @@
+
+void setup() {
+  Serial.begin(9600);
+  while (!Serial) {
+    ; // Wait for serial port to connect.
+  }
+  Serial.write("WebUSB FTW!");
+  Serial.flush();
+}
+
+void loop() {
+  // Nothing here for now.
+}
+{% endhighlight %}
+
+The third-party [WebUSB Arduino
+library](https://github.com/webusb/arduino/tree/gh-pages/library/WebUSB) used
+in the sample code above does basically two things:
+
+- The device acts as a WebUSB device enabling Chrome to read
+  the [landing page URL](https://github.com/webusb/arduino/blob/25e33acb0345fffa610c07aa20c898e2a3861859/library/WebUSB/WebUSB.cpp#L53)
+  and the [list of origins](https://github.com/webusb/arduino/blob/25e33acb0345fffa610c07aa20c898e2a3861859/library/WebUSB/WebUSB.cpp#L130) allowed to communicate with.
+- It exposes a WebUSB Serial API that you may use to override the default one.
+
+Let's focus back to the JavaScript code. Once we get the `device` picked by the
+user, `device.open` simply runs all platform-specific steps to start a session
+with the USB device. Then, we have to select an available USB Configuration
+with `device.selectConfiguration`. Remember that a Configuration specifies how
+the device is powered, its maximum power consumption and its number of
+interfaces. Talking about interfaces, we also need to request exclusive access
+with `device.claimInterface` since data can only be transfered to an interface
+or associated endpoints when the interface is claimed. Finally calling
+`device.controlTransferOut` is needed to set up the Arduino device with the
+appropriate commands to communicate through the WebUSB Serial API.
+
+From there, `device.transferIn` performs a Bulk Transfer onto the
+device to inform that the host is ready to receive bulk data. As soon as it
+gets it, the promise is fulfilled with a `result` object containing a
+[DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView)
+`data` that has to be parsed appropriately.
+
+For those who are familiar with USB, all of this should look pretty familiar.
+
+### I want moar
+
+The WebUSB API lets you interact with the all USB transfer/endpoint types:
+
+- CONTROL transfers, used to send or receive configuration or command
+  parameters to a USB device are handled with `controlTransferIn(setup,
+  length)` and `controlTransferOut(setup, data)`.
+- INTERRUPT transfers, used for small amount of time sensitive data are handled
+  with same methods as BULK transfers with `transferIn(endpointNumber, length)`
+  and `transferOut(endpointNumber, data)`.
+- ISOCHRONOUS transfers, used for streams of data, like video and sound are
+  handled with `isochronousTransferIn(endpointNumber, packetLengths)` and
+ `isochronousTransferOut(endpointNumber, data, packetLengths)`.
+- BULK transfers, used to transfer a large amount of non-time-sensitive data in
+  a reliable way are handled with `transferIn(endpointNumber, length)` and
+ `transferOut(endpointNumber, data)`.
+
```

Comment:
Maybe link to Mike Tsao's [WebLight](https://github.com/sowbug/weblight) project which provides a ground-up example of building a WebUSB device (not using an Arduino).